### PR TITLE
chore(KFLUXVNGD-128): Delete old crossplane namespace claims

### PR DESCRIPTION
**Pull request by hmariset**
Delete crossplane namespace claims periodically to prevent the compute provided.

Jira-Url: https://issues.redhat.com/browse/KFLUXVNGD-128

### CODE REVIEW

File: components/namespace-claim-cleaner/base/rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+metadata:
+  name: namespace-claim-cleaner
+rules:
+  - apiGroups:
+      - toolchain.dev.openshift.com
+    resources:
+      - namespace-claim
+    verbs:
+      - list
+      - delete
```
Comment (amisstea) [inline components/namespace-claim-cleaner/base/rbac.yaml:19]: This isn't a valid resource nor should this role/SA interact with anything related to the toolchain operators. Please remove this.

File: components/namespace-claim-cleaner/base/rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - apiGroups:
+      - ""
+    resources:
+      - namespaces
+    verbs:
+      - list
```
Comment (amisstea) [inline components/namespace-claim-cleaner/base/rbac.yaml:25]: We shouldn't need to list core namespaces. We need to be able to list and delete namespace in the `eaas.konflux-ci-dev` api group.

File: components/namespace-claim-cleaner/base/namespace-claim-configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+kind: ConfigMap
+metadata:
+  name: namespace-claim-configmap
+  namespace: namespace-claim-cleaner
+data:
+  namespace-claim-cleaner.sh: |
+    #!/bin/bash
+
+    # Get namespaces with label namespaces.eaas.konflux-ci-dev
+    LABEL="namespaces.eaas.konflux-ci-dev"
```
Comment (amisstea) [inline components/namespace-claim-cleaner/base/namespace-claim-configmap.yaml:11]: We can't list the namespaces generated by the claim. They're not available on the cluster where this code runs.

This is more complex that it needs to be. We want to list and delete the stale `namespaces.eaas.konflux-ci-dev` resources.

**Review by amisstea — CHANGES_REQUESTED** [review]
Why do we need another component for this? I was thinking this CronJob and its config could reside in `crossplane-control-plane` since it's a tightly coupled dependency.

File: components/crossplane-control-plane/base/cronjob.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  namespace: crossplane-control-plane
+spec:
+  schedule: "0 4 * * *" # every day at 4AM UTC
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          containers:
+            - name: namespace-claim-cleaner
+              image: bitnami/kubectl:latest
```
Comment (amisstea) [inline components/crossplane-control-plane/base/cronjob.yaml:14]: We shouldn't use `docker.io` due to security risks and rate limiting issues. Pulling the `latest` tag is also generally ill advised for production code - we should pull by a more specific tag and digest whenever possible. Please use the [appstudio-utils](https://quay.io/repository/konflux-ci/appstudio-utils?tab=tags) image instead which has `kubectl` and `jq`/`yq` baked in.

File: components/crossplane-control-plane/base/namespace-claim-configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+metadata:
+  name: namespace-claim-configmap
+  namespace: crossplane-control-plane
+data:
+  namespace-claim-cleaner.sh: |
+    #!/bin/bash
+
+    CURRENT_TIME=$(date +%s)
+    kubectl get namespaces.eaas.konflux-ci.dev -o json | jq -c '.items[]' | while read -r claim; do
+      CREATION_TIMESTAMP=$(echo $claim | jq -r '.metadata.creationTimestamp')
```
Comment (amisstea) [inline components/crossplane-control-plane/base/namespace-claim-configmap.yaml:12]: No pipeline required (untested but should work):
```suggestion
      CREATION_TIMESTAMP=$(jq -r '.metadata.creationTimestamp' <<< "$claim")
```

File: components/crossplane-control-plane/base/namespace-claim-configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      CREATION_TIME=$(date -d "$CREATION_TIMESTAMP" +%s)
+      AGE=$((CURRENT_TIME - CREATION_TIME))
+      if [ $AGE -gt (2*24*3600) ]; then
+        NAME=$(echo $claim | jq -r '.metadata.name')
```
Comment (amisstea) [inline components/crossplane-control-plane/base/namespace-claim-configmap.yaml:16]:
```suggestion
        NAME=$(jq -r '.metadata.name' <<< "$claim")
```

File: components/crossplane-control-plane/base/namespace-claim-configmap.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+data:
+  namespace-claim-cleaner.sh: |
+    #!/bin/bash
+
+    CURRENT_TIME=$(date +%s)
+    kubectl get namespaces.eaas.konflux-ci.dev -o json | jq -c '.items[]' | while read -r claim; do
+      CREATION_TIMESTAMP=$(echo $claim | jq -r '.metadata.creationTimestamp')
+      CREATION_TIME=$(date -d "$CREATION_TIMESTAMP" +%s)
+      AGE=$((CURRENT_TIME - CREATION_TIME))
+      if [ $AGE -gt (2*24*3600) ]; then
```
Comment (amisstea) [inline components/crossplane-control-plane/base/namespace-claim-configmap.yaml:15]: This is checking if the claim is older than 2 days, right? If so, why so long?

File: components/crossplane-control-plane/base/namespace-claim-configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: namespace-claim-configmap
+  namespace: crossplane-control-plane
+data:
+  namespace-claim-cleaner.sh: |
+    #!/bin/bash
```
Comment (amisstea) [inline components/crossplane-control-plane/base/namespace-claim-configmap.yaml:8]: Is this going to fail properly if any of the piped commands below fail? I suspect it won't without `set`ting some options.

**Review by amisstea — COMMENTED** [review]
@hmariset please include some evidence showing how this was tested in a dev cluster with it successfully cleaning up stale namespace claims while ignoring others.

File: components/crossplane-control-plane/base/kustomization.yaml [hunk 9]
```diff
@@ -1,6 +1,9 @@
 resources:
-- https://github.com/konflux-ci/crossplane-control-plane/crossplane?ref=5d6c42730c1c9f66b5d3567bdf04d587832ceac1
-- https://github.com/konflux-ci/crossplane-control-plane/config?ref=5d6c42730c1c9f66b5d3567bdf04d587832ceac1
+- https://github.com/konflux-ci/crossplane-control-plane/crossplane?ref=f61fa6c74ed6e7ac1a913c35612498157b37549d
+- https://github.com/konflux-ci/crossplane-control-plane/config?ref=f61fa6c74ed6e7ac1a913c35612498157b37549d
+- rbac.yaml
+- cronjob.yaml
+- namespace-claim-configmap.yaml
```
Comment (amisstea) [inline components/crossplane-control-plane/base/kustomization.yaml:6]: Nit: be consistent with filenames & resource names.
```suggestion
- configmap.yaml
```

**Comment by hmariset** [comment]
image
Successfully deleted "my-namespace" in "test-pvc".

**Comment by hmariset** [comment]
image
When there are no namespaces.eaas.konflux-ci.dev resources.

**Comment by hmariset** [comment]
To create sample `namespaces.eaas.konflux-ci.dev` object you can use the example from [`crossplane-control-plane`](https://github.com/konflux-ci/crossplane-control-plane/blob/main/examples/xnamespace/claim.yaml) repo.
`kubectl apply -f examples/xnamespace/claim.yaml`

File: components/crossplane-control-plane/base/namespace-claim-configmap.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: namespace-claim-configmap
```
[the remark this answers is not in the document]
Comment (hmariset) [reply components/crossplane-control-plane/base/namespace-claim-configmap.yaml]: Updated.

**Review by amisstea — APPROVED** [review]
/lgtm

**Pull request merged**